### ISSUE TITLE
Dead loop in Okio AsyncTimeout caused by a circle in LinkedList

### ISSUE THREAD

**Issue by yihuaqi** [square/okio#624]
We're using com.sqaureup.okio:okio:1.14.0 and com.squareup.okhttp3:okhttp:3.10.0.2.

During developing, we occasionally have the issue that all our network requests are not executing because RealBufferedSink cannot flush. We looked into the issue deeper, and found out that one of our thread is stuck at a dead loop in AsyncTimeout:
```
  private static synchronized void scheduleTimeout(
      AsyncTimeout node, long timeoutNanos, boolean hasDeadline) {
    // Start the watchdog thread and create the head node when the first timeout is scheduled.
    if (head == null) {
      head = new AsyncTimeout();
      new Watchdog().start();
    }

    long now = System.nanoTime();
    if (timeoutNanos != 0 && hasDeadline) {
      // Compute the earliest event; either timeout or deadline. Because nanoTime can wrap around,
      // Math.min() is undefined for absolute values, but meaningful for relative ones.
      node.timeoutAt = now + Math.min(timeoutNanos, node.deadlineNanoTime() - now);
    } else if (timeoutNanos != 0) {
      node.timeoutAt = now + timeoutNanos;
    } else if (hasDeadline) {
      node.timeoutAt = node.deadlineNanoTime();
    } else {
      throw new AssertionError();
    }

    // Insert the node in sorted order.
    long remainingNanos = node.remainingNanos(now);
    for (AsyncTimeout prev = head; true; prev = prev.next) {
      if (prev.next == null || remainingNanos < prev.next.remainingNanos(now)) {
        node.next = prev.next;
        prev.next = node;
        if (prev == head) {
          AsyncTimeout.class.notify(); // Wake up the watchdog when inserting at the front.
        }
        break;
      }
    }
  }
```
Then we set a break point in the for loop, and found out that prev.next is prev itself, meaning that there's a circle in LinkedList, which causes the dead loop.

We're not able to reproduce it consistently, but it happens more frequently when there're a lot of network requests running at the same time, thus we suspect it to be a rare race condition.

And this happens during development, so it's not the same issue as https://github.com/square/okio/issues/185 since no ProGuard is involved.

I'm not sure that if it is a known issue that has been fix in the later version. What we're going to do to work around this issue is that we will periodically check if there's a circle in that LinkedList and fix that by using reflection. If anyone has a better fix please suggest.

Thanks in advance.

**Comment by swankjesse** [comment]
Does your code ever read or write a stream from multiple threads concurrently? Even perhaps just closing from the wrong thread? That'd potentially trigger this.

**Comment by yihuaqi** [comment]
> Does your code ever read or write a stream from multiple threads concurrently? Even perhaps just closing from the wrong thread? That'd potentially trigger this.

Wouldn't that raise unbalanced enter/exit exception? We tried that during development and found that wouldn't work because enter()/exit()/close() ensures only one thread can operate on a stream at a time.

In order to avoid that, our code will read the stream from the same thread, wrap the bytes into several ByteArrayStreams, and then dispatch the ByteArrayStreams to different threads for further processing. Will that cause problem too?

**Comment by swankjesse** [comment]
Can you provide an executable test case to reproduce the problem?

**Comment by yihuaqi** [comment]
> Can you provide an executable test case to reproduce the problem?

No, I can't. This doesn't happens very frequently (about 3 times a week among our team of 30 members). Just by looking at the code, there might be a race condition between cancelScheduledTimeout() and scheduleTimeout() that could result in a circle. I tried to write a test case that calls enter()/exit() thousand of times in different threads with near deadlines, but cannot reproduce it.

An interesting thing is that every time this happened it's always the smallest circle (prev.next == prev), so it might be performance-wise acceptable to just add this check in that for loop? It's hard to say how much it would impact the users because there's no crash and user cannot give feedback at the time it happens.

**Comment by yihuaqi** [comment]
I wrote a background task to periodically check the circle, and found that nearly 10% of the users are suffering from the issue, and 3 times a day on average.

```
    private fun AsyncTimeout.checkAndFixLoop(): Boolean {
        val nodes = HashSet<AsyncTimeout>()
        var current = this
        while(true) {
            Logger.d(TAG, "checkAndFixLoop $current")
            // A convenient reflection util.
            val next: AsyncTimeout? = Reflect.on(current).get<AsyncTimeout>("next")
            if (nodes.contains(next)) {
                OkioDeadLoopEvent().sendEvent()
                // break the circle
                Reflect.on(current).set("next", null)
                return true
            }
            if (next == null) {
                return false
            }
            nodes.add(next)
            current = next
        }
    }
```

But our QA spent a day and still cannot reproduce it, so I'm wondering if my code also has some race condition with AsyncTimeout (during LinkedList sorting?). Will it be better if I only check that `this == this.next` in AsyncTimeout?

**Comment by realskill** [comment]
I have the same issue, is there a solution for this?

**Comment by swankjesse** [comment]
@realskill Can you provide an executable test case to reproduce the problem?

**Comment by realskill** [comment]
No, it's rare. I found a bug like this in JDK, https://bugs.openjdk.java.net/browse/jdk-8172578, but my version of jdk1.8.0_161 should have been fixed. The problem is in the response.body().string() method of okhttp3.10. Now I use the callTimeout method provided by okhttp3.12 to solve it (only connectTimeout,readTimeout and writeTimeout are available in 3.10).

**Comment by findepi** [comment]
I hit a deadlock today. I identified this is an infinite loop in `cancelScheduledTimeout` because the `head -> next -> next` list had a cycle -- in my case it had two elements, and second was pointing at itself.

There seems to be at least one condition where this is possible, see https://github.com/square/okio/pull/818
